src/Expo.php

```diff
@@ -142,68 +161,119 @@ public function getRecipients()
+
+        foreach ($messages as $index => $message) {
+            if (!($message instanceof ExpoMessage))
+                $messages[$index] = new ExpoMessage($message);
+        }
+
+        $this->messages = $messages;
 
         return $this;
     }
 
     /**
-     * Sets the recipients to send the message to
+     * Sets default recipients
      *
      * @param string|array $recipients
      * @throws InvalidTokensException
      * @throws ExpoException
      */
     public function to($recipients = null): self
     {
-        $tokens = null;
-
-        if (is_array($recipients) && count($recipients) > 0) {
-            $tokens = $recipients;
-        } elseif (is_string($recipients)) {
-            $tokens = [$recipients];
-        } else {
-            throw new InvalidTokensException(sprintf(
-                'Tokens must be a string or non empty array, %s given.',
-                gettype($tokens)
-            ));
-        }
-
-        $tokens = array_filter($tokens, function ($token) {
-            return Utils::isExpoPushToken($token);
-        });
-
-        if (count($tokens) === 0) {
-            throw new ExpoException('No valid expo tokens provided.');
-        }
-
-        $this->recipients = $tokens;
+        $this->recipients = Utils::validateTokens($recipients);
 
         return $this;
     }
 
     /**
-     * Send the message to the expo server
+     * Send the messages to the expo server
      *
      * @throws ExpoException
      */
     public function push(): ExpoResponse
     {
-        if (is_null($this->message) || is_null($this->recipients)) {
-            throw new ExpoException('You must have a message and recipients to push');
+        if (empty($this->messages)) {
+            throw new ExpoException('You must have messages to push');
         }
 
-        $messages = array_map(function ($recipient) {
-            return $this->message->toArray() + ['to' => $recipient];
-        }, $this->recipients);
+        $messages = [];
+
+        /**
+         * When response ticket has DeviceNotRegistered it has no indication which push token produced this error,
+         * however it is known that the order of messages and response tickets are the same.
+         * So the only way to keep track of invalid tokens is by their indexes.
+         * For this to work we need to flatten messages' recipients, e.g.
+         * {
+         *     'to' => ['token1', 'token2'],
+         *     'body' => 'Message body',
+         * }
+         * would turn into
+         * [
+         *     {
+         *         'to' => 'token1',
+         *         'body' => 'Message body',
+         *     },
+         *     {
+         *         'to' => 'token2',
+         *         'body' => 'Message body',
+         *     },
+         * ]
+         */
+        foreach ($this->messages as $message) {
+            $message = $message->toArray();
+            foreach (Utils::arrayWrap($message['to'] ?? $this->recipients) as $token) {
```

Comment:
@ctwillie 
Well, from logic perspective, `sending messages to 0 recipients` is the same as `sending no messages`, thus skipping messages with no recipients would be expected.

Expo, being the end-service, cannot allow such things because it has to give response to all requests (including those containing messages with no recipients). But services/APIs/whatever that are sending requests to Expo have control over such things and can just not send the request for such messages (or not include them in request), which would produce result that is very close to what you would logically expect.